src/base/bittorrent/torrenthandle.cpp

```diff
@@ -1298,15 +1298,18 @@ void TorrentHandle::toggleSequentialDownload()
     setSequentialDownload(!isSequentialDownload());
 }
 
-void TorrentHandle::setFirstLastPiecePriority(const bool enabled)
+void TorrentHandle::setFirstLastPiecePriority(const bool enabled, const QVector<int> &updatedFilePrio)
 {
+    // Download first and last pieces first for every file in the torrent
+
     if (!hasMetadata()) {
         m_needsToSetFirstLastPiecePriority = enabled;
         return;
     }
 
-    // Download first and last pieces first for every file in the torrent
-    const std::vector<int> filePriorities = nativeHandle().file_priorities();
+    // Updating file priorities is an async operation in libtorrent, when we just updated it and immediately query it
+    // we might get the old/wrong values, so we rely on `updatedFilePrio` in this case.
+    const std::vector<int> filePriorities = !updatedFilePrio.isEmpty() ? updatedFilePrio.toStdVector() : nativeHandle().file_priorities();
```

Comment:
I under the impression that RVO could kick in here... ?

Reply:
RVO is essentially when a function constructs its return value directly in the location of the callers variable. in this case, the `QVector` returns a copy of itself, and it will have to be copied since you still have the `QVector` object left at the end.

Reply:
>it seems a bit excessive to copy the whole vector here. 

I suppose it is constructing with a rvalue, then move semantics could be in effect here?

Reply:
it will make a copy either way, because you end up with two copies at the end. RVO and move semantics may kick in to avoid *additional* copies. With my first comment, I didn't realise you were actually converting from `QVector`, I thought it was a copy just because `nativeHandle().file_priorities()` returns a temporary. But since you're converting the types, there probably isn't a way around the copy

Reply:
@Chocobo1, maybe we should use STD types where there are no points of contact with Qt to avoid unnecessary converting?
I.e. if some container are created in application code and it isn't supposed to be passed to Qt code we can use std type for it. Otherwise, if some container are coming from Qt or should be passed to Qt then we can use Qt type.
So the question is "How file priorities are used across application code?".

Reply:
>@Chocobo1, maybe we should use STD types where there are no points of contact with Qt to avoid unnecessary converting?

No objections from me.

>So the question is "How file priorities are used across application code?".

When interacting with libtorrent, it must be std container; when dealing with the GUI, currently it is Qt container.
Talking about `setFirstLastPiecePriorityImpl()`, do you agree the second parameter should be std::vector instead?

Reply:
>When interacting with libtorrent, it must be std container; when dealing with the GUI, currently it is Qt container.

You must have misunderstood me. Well, I checked it out. Look.
First time it is created in `TorrentContentModel::getFilePriorities()` and it is never passed into Qt. So we can use std container for file priorities everywhere, starting from its creation.

Reply:
Currently it might look a bit out of place seeing a std container there, one could wonder why it doesn't use a Qt container, also following this logic we could end up using std containers almost everywhere.

If everyone agrees with it, then why not. However I wish to defer this change later, it isn't really related to the scope of this PR.

Reply:
>However I wish to defer this change later, it isn't really related to the scope of this PR.

Ok.